File: ABC/ABC190/E.py
import sys
from collections import deque

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


def solve():
    N, M = map(int, rl().split())
    graph = [[] for _ in range(N)]
    for _ in range(M):
        ai, bi = map(lambda n: int(n) - 1, rl().split())
        graph[ai].append(bi)
        graph[bi].append(ai)
    K = int(rl())
    C = list(map(lambda n: int(n) - 1, rl().split()))
    
    c_set = set(C)
    c_to_idx = {ci: i for i, ci in enumerate(C)}
    INF = 10 ** 18
    dist = [[INF] * K for _ in range(K)]
    for i in range(K):
        que = deque([(-1, C[i], 0)])
        searched = [False] * N
        searched[C[i]] = True
        while que:
            parent, cur, d = que.popleft()
            for child in graph[cur]:
                if child == parent or searched[child]:
                    continue
                if child in c_set:
                    dist[i][c_to_idx[child]] = d + 1
                searched[child] = True
                que.append((cur, child, d + 1))
    
    dp = [[INF] * K for _ in range(2 ** K)]
    for i in range(K):
        dp[1 << i][i] = 0
    
    for s in range(1, 2 ** K):
        for i in range(K):
            if s >> i & 1:
                continue
            ns = s | (1 << i)
            for j in range(K):
                if i == j:
                    continue
                dp[ns][i] = min(dp[ns][i], dp[s][j] + dist[j][i])
    
    ans = min(dp[2 ** K - 1])
    if ans == INF:
        print(-1)
    else:
        print(ans + 1)


if __name__ == '__main__':
    solve()

File: ABC/ABC190/E.nim
import algorithm, bitops, deques, heapqueue, math, sets, sequtils, strformat, strutils, sugar, tables

proc input*(): string {.inline.} = stdin.readLine
proc inputs*(): seq[string] {.inline.} = stdin.readLine.split
proc inputInt*(): int {.inline.} = stdin.readLine.parseInt
proc inputInts*(): seq[int] {.inline.} = stdin.readLine.split.map(parseInt)
proc chmax*[T: SomeNumber](num0: var T, num1: T) {.inline.} = num0 = max(num0, num1)
proc chmin*[T: SomeNumber](num0: var T, num1: T) {.inline.} = num0 = min(num0, num1)
proc `%=`*[T: SomeInteger](num0: var T, num1: T) {.inline.} = num0 = floorMod(num0, num1)

when isMainModule:
  var N, M: int
  (N, M) = inputInts()
  var
    graph = newSeqWith(N, newSeq[int]())
    ai, bi: int
  for _ in 0..<M:
    (ai, bi) = inputInts()
    dec ai; dec bi
    graph[ai].add(bi)
    graph[bi].add(ai)
  let K = inputInt()
  var C = inputInts()
  C.applyIt(it - 1)

  var cToIdx = initTable[int, int]()
  for i, ci in C:
    cToIdx[ci] = i
  
  var
    dist: array[17, array[17, int]]
    searched: array[10^5, bool]
    cSet = C.toHashSet()
  const INF = 10^18
  for i in 0..<K:
    dist[i].fill(INF)
  
  for i, ci in C:
    var que = initDeque[(int, int)]()
    que.addLast((ci, 0))
    searched.fill(false)
    searched[ci] = true

    var cur, d: int
    while 0 < que.len:
      (cur, d) = que.popFirst()
      for to in graph[cur]:
        if searched[to]:
          continue
        searched[to] = true
        que.addLast((to, d + 1))
        if to in cSet:
          dist[i][cToIdx[to]] = d + 1
  
  var dp: array[2^17, array[17, int]]
  for i in 0..<2^K:
    dp[i].fill(INF)
  for i in 0..<K:
    dp[1 shl i][i] = 1

  for S in 1..<2^K:
    for i in 0..<K:
      if bitand(S shr i, 1) == 1:
        continue
      let nS = bitor(S, 1 shl i)
      for j in 0..<K:
        if i == j:
          continue
        dp[nS][i].chmin(dp[S][j] + dist[j][i])
  
  let ans = min(dp[2^K - 1])
  echo if ans != INF: ans else: -1
